packages/fuel-indexer-graphql/src/query.rs

```diff
@@ -0,0 +1,684 @@
+    GraphqlError, GraphqlResult,
+};
+
+/// Contains information about a successfully-parsed user operation.
+#[derive(Debug)]
+pub struct ParsedOperation {
+    pub name: Option<String>,
+    pub selections: Vec<SelectionNode>,
+    pub ty: OperationType,
+    pub variable_definitions: Vec<Positioned<VariableDefinition>>,
+    pub directives: Vec<Positioned<Directive>>,
+}
+
+/// Contains necessary information for generating joins between database tables.
+#[derive(Debug, Default)]
+pub struct DependencyGraph {
+    pub table_node_idx_map: HashMap<String, NodeIndex>,
+    pub graph: Graph<String, (String, String)>,
+}
+
+impl DependencyGraph {
+    /// Add a new node to dependency graph.
+    fn add_node(&mut self, table: String) -> NodeIndex {
+        if let Some(existing_node_idx) = self.table_node_idx_map.get(&table) {
+            *existing_node_idx
+        } else {
+            let new_node_idx = self.graph.add_node(table.clone());
+            self.table_node_idx_map.insert(table, new_node_idx);
+            new_node_idx
+        }
+    }
+
+    /// Add an edge between two existing nodes.
+    fn add_edge(
+        &mut self,
+        parent: NodeIndex,
+        child: NodeIndex,
+        referring_field: String,
+        foreign_key_field: String,
+    ) {
+        self.graph
+            .add_edge(parent, child, (referring_field, foreign_key_field));
+    }
+
+    /// Returns database joins in topologically sorted order.
+    fn get_sorted_joins(&self) -> GraphqlResult<Vec<String>> {
+        let toposorted_nodes =
+            if let Ok(sorted) = petgraph::algo::toposort(&self.graph, None) {
+                sorted
+            } else {
+                return Err(GraphqlError::QueryError(
+                    "Cannot have cycles in query".to_string(),
+                ));
+            };
+
+        let mut joins: Vec<String> = vec![];
+        let mut seen = vec![false; self.graph.node_count()];
+
+        let mut stack = VecDeque::new();
+        stack.push_back(toposorted_nodes[0]);
+
+        while let Some(node_idx) = stack.pop_front() {
+            if seen[node_idx.index()] {
+                continue;
+            }
+
+            let mut neighbors = self
+                .graph
+                .neighbors_directed(node_idx, petgraph::Direction::Outgoing)
+                .detach();
+
+            while let Some(e) = neighbors.next_edge(&self.graph) {
+                if let (
+                    Some((parent_node, child_node)),
+                    Some((referring_field, fk_field)),
+                ) = (self.graph.edge_endpoints(e), self.graph.edge_weight(e))
+                {
+                    if let (Some(parent_table), Some(child_table)) = (
+                        self.graph.node_weight(parent_node),
+                        self.graph.node_weight(child_node),
+                    ) {
+                        let join = format!(
+                            "INNER JOIN {} ON {}.{} = {}.{}",
```

Comment:
Can we extract this into some struct instead of the string hacking? This would help make things a bit more maintainable as you'd know what are the components of this `InnerJoin` and you could always `impl Display for InnerJoin` in order to get the SQL fragment (or `impl SqlFragment for InnerJoin`)